### PR TITLE
Add SharedGlobalKey

### PR DESCRIPTION
**Pull request by simonmaxen-da**
This PR is part of a program of work to support shared contract keys. The first phase of this is making packageId optional inside contract keys, see https://github.com/DACH-NY/canton/issues/14486.

This PR introduces `SharedGlobalKey` that is the target structure for `GlobalKey` the difference being that the `packageId` is optional inside `SharedGlobalKey`.

In many cases the `packageId`, although not in the contract key will be available from the context in which the code is running. When it comes to error handling this is not the case.

An example of a problem is when the `GrpcErrorParser` gets passed a `SharedContractKey`. In this situation no `packageId` is available. For situations where the a conversion from value to SValue is needed for _error reporting_ a `LossyValueTranslator` has been added.

There are alternatives to doing this once we really have LF support for shared contract keys, see:
https://github.com/digital-asset/daml/issues/17646

### CODE REVIEW

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala [hunk 1]
```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package engine
+package preprocessing
+
+import com.daml.lf.data.Ref.IdString
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala:8]:
```suggestion
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+              ),
+              TApp(TTyCon(lossyConversionIdentifier), vt.ty),
+            )
+
+          case ValueRecord(mTyCon, sourceElements) =>
+            val flatNames = sourceElements.iterator.flatMap(_._1).toList
+            val names = flatNames.length match {
+              case 0 =>
+                (1 to sourceElements.length).map(i =>
+                  IdString.Name.assertFromString(s"$lossyConversionFieldPrefix$i")
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala:144]:
```suggestion
                  Ref.Name.assertFromString(s"$lossyConversionFieldPrefix$i")
```

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/GlobalKey.scala [hunk 3]
```diff
@@ -25,12 +27,67 @@ final class GlobalKey private (
...
   def qualifiedName: Ref.QualifiedName = templateId.qualifiedName
+  def shared: SharedGlobalKey =
+    SharedGlobalKey(Some(templateId.packageId), templateId.qualifiedName, key, hash)
 
   override def hashCode(): Int = hash.hashCode()
 
   override def toString: String = s"GlobalKey($templateId, $key)"
 }
 
+// The target state for GlobalKey once all call sites support SharedGlobalKey (#14486)
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/GlobalKey.scala:38]: Can you create an issue about removing the `pakageId` in contract key and link it here 
```suggestion
// TODO: https://github.com/digital-asset/daml/issues/XXXX
//   The target state for GlobalKey once all call sites support SharedGlobalKey (#14486)
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+package preprocessing
+
+import com.daml.lf.data.Ref.IdString
+import com.daml.lf.data._
+import com.daml.lf.language.Ast._
+import com.daml.lf.speedy.{ArrayList, SValue}
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value._
+
+/** This translator is only intended for use where a SValue is needed for error reporting but
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala:15]: can you link the issue I ask you to create lower.

Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala:15]: Could we move this to Daml script to be sure nobody else use it.

**Review by remyhaemmerle-da — APPROVED** [review]
Looks reasonable.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala (hunk 4 above)
Comment (simonmaxen-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala:15]: Following a conversation with @remyhaemmerle-da it was decided to leave it closer to `ValueTranslator`.

Comment (simonmaxen-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/LossyValueTranslator.scala:15]: Updated comment to reference the daml issue.

**Pull request merged**